### PR TITLE
add logging mechanism

### PR DESCRIPTION
**Pull request by lplewa**
TODO:
- [ ] Documentation
- [ ] Windows

### CODE REVIEW

File: src/utils/utils_log.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
+#define _GNU_SOURCE 1
```
Comment (KFilipek) [inline src/utils/utils_log.c:1]: Please, add the missing licences

File: test/test_utils_log.cpp [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+#define getenv(X) strstr(X, "UMF_LOG") ? (char *)env_variable : getenv(X)
+#include "utils/utils_log.c"
+#undef getenv
+#undef fopen
+#undef fputs
+#undef fflush
+}
+using umf_test::test;
+
+TEST_F(test, matchEnvVar) {
```
Comment (igchor) [inline test/test_utils_log.cpp:56]: Could you add some tests for empty string and string with `;` at the beginning?

File: test/test_utils_log.cpp [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                        helper_log_init(envVar.c_str());
+                        helper_checkConfig(&b, &loggerConfig);
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Unfortunately this must be a macro to pass variadic argument
```
Comment (igchor) [inline test/test_utils_log.cpp:216]: Not necessarily, you can probably do:

```
template <typename... Args>
void helper_test_log(Args... args) {
put_count = 0;                                                        
        fflush_count = 0;                                                      
        utils_log(args...);                                                
        EXPECT_EQ(fput_count, expect_fput_count);                              
        EXPECT_EQ(fflush_count, expect_fflush_count);
}
```

File: test/test_utils_log.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+            helper_test_log((LogLevel)j, "example log");
+        }
+    }
+}
+
+TEST_F(test, long_log) {
+    expect_fput_count = 1;
+    expect_fflush_count = 1;
+    loggerConfig = {0, 0, LOG_DEBUG, LOG_DEBUG, stderr};
+    expected_message = "[DEBUG UMF] " LONG_STRING_2047 "\n";
```
Comment (igchor) [inline test/test_utils_log.cpp:293]: nit: instead of having all those LONG_STRING defines you can just do:
```
"[DEBUG UMF] " + std::string(2047, 'x') +  "\n"
``` 
or just make a function that returns std::string of specified length if you need to have specific content

Comment (lplewa) [reply test/test_utils_log.cpp]: Done

File: test/test_utils_log.cpp (hunk 3 above)
Comment (lplewa) [reply test/test_utils_log.cpp]: Right, we can use cpp magic :rofl:

File: test/test_utils_log.cpp (hunk 2 above)
Comment (lplewa) [reply test/test_utils_log.cpp:56]: Done

File: src/utils/utils_log.c [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+#include "utils_common.h"
+#include "utils_log.h"
+
+typedef struct {
+    int timestamp;
+    int pid;
+    LogLevel level;
+    LogLevel flushLevel;
+    FILE *output;
+} LoggerConfig;
```
Comment (bratpiorka) [inline src/utils/utils_log.c:30]: see how the struct definition looks like in UMF -rename to umf_logger_config_t etc

File: src/utils/utils_log.c [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+LoggerConfig loggerConfig = {0, 0, LOG_ERROR, LOG_ERROR, NULL};
+
+static const char *levelToStr(LogLevel l) {
```
Comment (bratpiorka) [inline src/utils/utils_log.c:34]: level_to_str ? please do not mix styles

File: src/utils/utils_log.c [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        return "ERROR";
+    case LOG_INFO:
+        return "INFO";
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+    }
+}
+static const char *matchEnvVar(const char *envVar, const char *prefix,
```
Comment (bratpiorka) [inline src/utils/utils_log.c:48]: add empty line before static

File: src/utils/utils_log.c [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    va_list args;
+    va_start(args, format);
+    int ret = vsnprintf(buffer, sizeof(buffer), format, args);
+    const char *overflow = "";
+    if (ret >= (intptr_t)sizeof(buffer)) {
+        //TODO: alloc bigger buffer with base alloc
+        overflow = "[truncated...]";
+    }
+    va_end(args);
+    char header[256];
```
Comment (bratpiorka) [inline src/utils/utils_log.c:106]: please add a #define for this and all magic numbers and move these #defines to the top of the file

File: src/utils/utils_log.c [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+    if (loggerConfig.pid) {
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
+        h_pos += snprintf(h_pos, sizeof(header) - (h_pos - header),
+                          "PID:%-6lu TID:%-6lu ", (unsigned long)pid,
+                          (unsigned long)tid);
+    }
+
+    char logLine[2560];
```
Comment (bratpiorka) [inline src/utils/utils_log.c:130]: 2560? please #define this

File: src/utils/utils_log.c [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+             levelToStr(level), buffer, overflow);
+
+    fputs(logLine, loggerConfig.output);
+
+    if (level >= loggerConfig.flushLevel) {
+        fflush(loggerConfig.output);
+    }
+}
+
+#define MAX_FILE_PATH 256
```
Comment (bratpiorka) [inline src/utils/utils_log.c:141]: are there any os-specific functions to get this?

File: src/utils/utils_log.c [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+void log_init(void) {
+    char envVar[2048];
```
Comment (bratpiorka) [inline src/utils/utils_log.c:143]: #define this

File: src/utils/utils_log.c [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    if (util_env_var("UMF_LOG", envVar, sizeof(envVar)) < 0) {
+        return;
+    }
+
+    if (matchEnvVar(envVar, "timestamp:yes", NULL)) {
+        loggerConfig.timestamp = 1;
+    } else if (matchEnvVar(envVar, "timestamp:no", NULL)) {
+        loggerConfig.timestamp = 0;
+    }
```
Comment (bratpiorka) [inline src/utils/utils_log.c:152]: else? Shouldn't we print something in case of timestamp:xyz?

File: src/utils/utils_log.c [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        char file[MAX_FILE_PATH + 1];
+        size_t len = 0;
+        if (argEnd) {
+            len = argEnd - arg;
+        } else {
+            len = strlen(arg);
+        }
+        if (len <= MAX_FILE_PATH) {
+#ifdef _WIN32
+            strncpy_s(file, len, arg, len);
```
Comment (bratpiorka) [inline src/utils/utils_log.c:196]: maybe just memcpy? since we know the len

File: src/utils/utils_log.c [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+#ifdef _WIN32
+    DWORD pid = GetCurrentProcessId();
+    DWORD tid = GetCurrentThreadId();
+#else
+    pid_t pid = getpid();
+    pid_t tid = gettid();
+#endif
+
+    char buffer[2048];
```
Comment (bratpiorka) [inline src/utils/utils_log.c:96]: make it 8K :)

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: src/utils/CMakeLists.txt [hunk 15]
```diff
@@ -5,6 +5,10 @@
 include(${UMF_CMAKE_SOURCE_DIR}/cmake/helpers.cmake)
 include(FindThreads)
 
+set(UMF_UTILS_SOURCES
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/CMakeLists.txt:8]: Done

File: src/utils/utils_log.c [hunk 16]
```diff
@@ -0,0 +1,204 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: src/utils/utils_log.h [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+#ifndef UMF_LOG_H
+#define UMF_LOG_H 1
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum { LOG_DEBUG, LOG_INFO, LOG_WARNING, LOG_ERROR } LogLevel;
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/utils_log.h]: kindof done - i changed it to match prefix utils_log

File: src/utils/utils_log.c (hunk 5 above)
Comment (lplewa) [reply src/utils/utils_log.c]: dito

File: src/utils/utils_log.c (hunk 7 above)
Comment (lplewa) [reply src/utils/utils_log.c]: This is how clang format formatted code

File: src/utils/utils_log.c (hunk 8 above)
Comment (lplewa) [reply src/utils/utils_log.c]: i added those defines. This is matter of codding style, but i proffer keep defines which are used only in the single function next to it definition. It's improves code readability, when you can see define and it's usage on single screen.

File: src/utils/utils_log.c (hunk 10 above)
Comment (lplewa) [reply src/utils/utils_log.c:148]: maybe - but not compile time. Also it's more our limitation, so we can handle this easier (and without allocations).

File: src/utils/utils_log.c (hunk 11 above)
Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: src/utils/utils_log.c (hunk 14 above)
Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: src/utils/utils_log.c [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#define _GNU_SOURCE 1
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/utils_log.c:10]: imho not needed, but done

File: src/utils/utils_log.c [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+    case LOG_INFO:
+        return "INFO";
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+        return "";
+    }
+}
+static const char *match_env_var(const char *envVar, const char *prefix,
```
Comment (bratpiorka) [inline src/utils/utils_log.c:51]: please add an empty line before function definition

File: src/utils/utils_log.c (hunk 5 above)
Comment (bratpiorka) [reply src/utils/utils_log.c]: I don't understand why we can't have one naming convention. It should be called "utils_log_config_t". I know we should clean this up in many existing places, but at least we shouldn't make it worse.

File: src/utils/utils_log.c (hunk 10 above)
Comment (bratpiorka) [reply src/utils/utils_log.c:156]: Please move all definitions at the top of the file. This is the first place were I look for them. Otherwise it requires ctrl+f to find the values.

File: src/utils/utils_log.c (hunk 12 above)
Comment (bratpiorka) [reply src/utils/utils_log.c:169]: ok but if user enables the logging somehow (set to output to e.g. stdout) and use the bad level then the logs would be empty and even without an info that something is wrong. What do you think about this scenario?

File: src/utils/utils_log.c [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+        size_t len = 0;
+        if (argEnd) {
+            len = argEnd - arg;
+        } else {
+            len = strlen(arg);
+        }
+        if (len <= MAX_FILE_PATH) {
+            memcpy(file, arg, len);
+            file[len] = '\0';
+#ifdef _WIN32
```
Comment (bratpiorka) [inline src/utils/utils_log.c:214]: I saw similar code in ubench_fopen - as this is the second time we have to deal with fopen - could you move the ubench version to the utils and use it here?

File: src/utils/utils_log.c [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+#define MAX_FILE_PATH 256
+#define MAX_ENV_LEN 2048
+void utils_log_init(void) {
+    char envVar[MAX_ENV_LEN];
+
+    if (util_env_var("UMF_LOG", envVar, sizeof(envVar)) < 0) {
+        return;
+    }
+
```
Comment (bratpiorka) [inline src/utils/utils_log.c:164]: I think we should have "defaults" that would be eventually overwritten - what do you think?

File: src/utils/utils_log.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/utils_log.c:10]: It must be before includes (at is affects included files)

File: src/utils/utils_log.c (hunk 21 above)
Comment (lplewa) [reply src/utils/utils_log.c:164]: check utils_log.c:34

This question, shows why things should be where they are used (instead of the begging of the file) (ref: to discussion about define positions)

File: src/utils/utils_log.c (hunk 19 above)
Comment (lplewa) [reply src/utils/utils_log.c:51]: Line added- 

And #define _GNU_SOURCE 1 affects others headers - it is common practice to define it as fist.

File: src/utils/utils_log.c (hunk 12 above)
Comment (lplewa) [reply src/utils/utils_log.c:205]: Then check what they did wrong ;)

Anyway - i see a point - please check updated version of the code

File: src/utils/utils_log.c [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+    default:
+        ASSERT(0);
+        return "";
+    }
+}
+
+static const char *match_env_var(const char *envVar, const char *prefix,
+                                 const char **extraArg) {
+    const char *found = strstr(envVar, prefix);
+    // ensure that found string is a first on list or it's separated semicolon
```
Comment (lukaszstolarczuk) [inline src/utils/utils_log.c:60]: did you mean?
`ensure that found string is first on the list or it's a seperating semicolon`

File: src/utils/utils_log.c [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+    if (!found) {
+        return NULL;
+    }
+    // if found string is not at the beginning of envVar ensure if it's preceded by ';'
```
Comment (lukaszstolarczuk) [inline src/utils/utils_log.c:64]: drop the second `if`

File: src/utils/utils_log.c [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+    if (found != envVar && *(found - 1) != ';') {
+        return NULL;
+    }
+
+    const char *endFound = found + strlen(prefix);
+    if (!extraArg) {
+        // if there is no argument matched string should end with ';' or '\0'
```
Comment (lukaszstolarczuk) [inline src/utils/utils_log.c:71]: `argument,` (missing coma)

File: src/utils/utils_log.c [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
+        h_pos += strftime(h_pos, sizeof(header) - (h_pos - header),
+                          "%Y-%m-%dT%H:%M:%S ", &tm_info);
+    }
+
+    if (loggerConfig.pid) {
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
```
Comment (lukaszstolarczuk) [inline src/utils/utils_log.c:140]: .

File: test/test_utils_log.cpp [hunk 26]
```diff
@@ -0,0 +1,380 @@
+// Copyright (C) 2024 Intel Corporation
```
Comment (lukaszstolarczuk) [inline test/test_utils_log.cpp:1]: pls drop `test_` from the name of the file

File: test/test_utils_log.cpp [hunk 27]
```diff
@@ -0,0 +1,380 @@
...
+#undef fopen
+#undef fputs
+#undef fflush
+}
+using umf_test::test;
+
+TEST_F(test, match_env_var) {
+    EXPECT_FALSE(match_env_var("", "test1", 0));
+
+    EXPECT_TRUE(match_env_var("test1;test2;test3;test4", "test1", 0));
```
Comment (lukaszstolarczuk) [inline test/test_utils_log.cpp:80]: pls add a test with `;` at the end

File: test/test_utils_log.cpp [hunk 28]
```diff
@@ -0,0 +1,380 @@
...
+    EXPECT_EQ(fput_count, expect_fput_count);
+    EXPECT_EQ(fflush_count, expect_fflush_count);
+
+    expected_message = "[ERROR UMF] example log\n";
+    fput_count = 0;
+    fflush_count = 0;
+    ERR("example log");
+    EXPECT_EQ(fput_count, expect_fput_count);
+    EXPECT_EQ(fflush_count, expect_fflush_count);
+}
```
Comment (lukaszstolarczuk) [inline test/test_utils_log.cpp:380]: can you pls add a test with a little longer logging, e.g.: log, log, log, flush, log, log, ... (especially with various log and flush levels)

File: test/test_utils_log.cpp [hunk 29]
```diff
@@ -0,0 +1,380 @@
...
+            } else {
+                expect_fflush_count = 1;
+            }
+            expected_message = "[" + helper_log_str(j) + " UMF] example log\n";
+            helper_test_log((utils_log_level_t)j, "example log");
+        }
+    }
+}
+
+TEST_F(test, long_log) {
```
Comment (lukaszstolarczuk) [inline test/test_utils_log.cpp:316]: ,can you run this test (and the next 3 tests) for various levels, not only DEBUG...?

File: src/utils/utils_log.c [hunk 30]
```diff
@@ -0,0 +1,237 @@
...
+static const char *bool_to_str(int b) { return b ? "yes" : "no"; }
+
+void utils_log_init(void) {
+    char envVar[MAX_ENV_LEN];
+
+    if (util_env_var("UMF_LOG", envVar, sizeof(envVar)) < 0) {
+        return;
+    }
+    const char *arg;
+    if (match_env_var(envVar, "output:stdout", NULL)) {
```
Comment (bratpiorka) [inline src/utils/utils_log.c:169]: please add documentation

File: src/utils/utils_log.c (hunk 20 above)
Comment (lplewa) [reply src/utils/utils_log.c]: ubench.h is an external dependency with a different license - we should not refactor such code without serious reason.

File: src/utils/utils_log.c (hunk 22 above)
Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: src/utils/utils_log.c (hunk 23 above)
Comment (lplewa) [reply src/utils/utils_log.c]: Done.

File: src/utils/utils_log.c (hunk 25 above)
Comment (lplewa) [reply src/utils/utils_log.c:146]: done

File: test/test_utils_log.cpp (hunk 26 above)
Comment (lplewa) [reply test/test_utils_log.cpp:1]: some files has test prefix, some doesn't - what is a rule here ( btw. I prefer names with test_prefix so you can distinguish it from file with implementation, without checking directories)

File: test/test_utils_log.cpp (hunk 27 above)
Comment (lplewa) [reply test/test_utils_log.cpp:86]: Done.

File: test/test_utils_log.cpp (hunk 28 above)
Comment (lplewa) [reply test/test_utils_log.cpp:386]: This is whitebox unit test  - we know that logger is stateless, so there is no need to test sequences.

File: test/test_utils_log.cpp (hunk 29 above)
Comment (lplewa) [reply test/test_utils_log.cpp:322]: I kept this test "unit" we already tested that log level works - there is no need duplicate this tests and make those tests bigger.

File: src/utils/utils_log.c [hunk 31]
```diff
@@ -0,0 +1,237 @@
...
+        time_t now = time(NULL);
+        struct tm tm_info;
+#ifdef _WIN32
+        localtime_s(&tm_info, &now);
+#else
+        localtime_r(&now, &tm_info);
+#endif
+
+        ASSERT((intptr_t)sizeof(header) >
+               (h_pos - header)); //should never happen
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: test/test_utils_log.cpp (hunk 26 above)
Comment (bratpiorka) [reply test/test_utils_log.cpp:1]: we have a mess in tests files - IMO we could leave it and do the cleanup in a separate PR

File: src/utils/utils_log.c (hunk 30 above)
Comment (lplewa) [reply src/utils/utils_log.c:174]: done

File: scripts/docs_config/introduction.rst [hunk 32]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+Logging in UMF is handled by logger. There are several levels of logging: *debug*, *info*, *warning*, and *error*.
+The level of logging determines what messages will be printed, ie. the level set to *warning* means all messages at levels *warning* and *error* will be printed.
+
+By default, there is a guarantee that *error* messages are flushed immediately. One can change this behavior to flush on lower-level messages.
+
+Loggers redirect messages to *stdout*, *stderr*, or a file
+
+By default, no messages are printed. To enable logger you have to set **UMF_LOG** environment variable which have following syntax for setting logger options:
+
+  "[level:debug|info|warning|error];[flush:<debug|info|warning|error>];[output:stdout|stderr|file=<path>];[timestamp:yes|no][pid:yes|no]"
```
Comment (PatKamin) [inline scripts/docs_config/introduction.rst:110]: There is a missing `;` between timestamp and pid

File: scripts/docs_config/introduction.rst [hunk 33]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+  * flush - a flush level, meaning that messages at this level and above are guaranteed to be flushed immediately,
+            possible values are the same as above,
+  * output - indicates where messages should be printed,
+             possible values are: *stdout*, *stderr* and *file*,
+             when providing a *file* output option, a *<path>* is required
+  * timestamp - add a timestamp to log message
+  * pid - add thread and process ids to log message
+
+  .. note::
+    For output to file, a path to the file have to be provided after a equals sign, like in the example above. The path has to exist, file will be created if not existing. Path to the file should be not longer then 255 characters long.
```
Comment (PatKamin) [inline scripts/docs_config/introduction.rst:123]: an equals

File: scripts/docs_config/introduction.rst [hunk 34]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+    output paramiter is required, all others logger options are optional. The defaults are set when options are not provided in the environment variable.
```
Comment (PatKamin) [inline scripts/docs_config/introduction.rst:124]: all other?

Comment (PatKamin) [inline scripts/docs_config/introduction.rst:124]: The output parameter

File: src/utils/utils_log.c [hunk 35]
```diff
@@ -0,0 +1,252 @@
...
+        h_pos += snprintf(h_pos, sizeof(header) - (h_pos - header),
+                          "PID:%-6lu TID:%-6lu ", (unsigned long)pid,
+                          (unsigned long)tid);
+    }
+
+    // We take twice header size here to ensure that
+    // we have space for log level and overflow string
+    // otherwise -Wformat-truncation might be thrown by compiler
+    char logLine[LOG_MAX + LOG_HEADER * 2];
+    snprintf(logLine, sizeof(logLine), "%s%-5s UMF] %s%s\n", header,
```
Comment (PatKamin) [inline src/utils/utils_log.c:154]: nit: `[` could be extracted from the header and placed directly here, easier to read.

File: src/utils/utils_log.c [hunk 36]
```diff
@@ -0,0 +1,252 @@
...
+    } else if (match_env_var(envVar, "flush:info", NULL)) {
+        loggerConfig.flushLevel = LOG_INFO;
+    } else if (match_env_var(envVar, "flush:warning", NULL)) {
+        loggerConfig.flushLevel = LOG_WARNING;
+    } else if (match_env_var(envVar, "flush:error", NULL)) {
+        loggerConfig.flushLevel = LOG_ERROR;
+    }
+
+    LOG_INFO(
+        "Logger enabled level: %s, flush: %s, pid: %s, timestamp: %s",
```
Comment (PatKamin) [inline src/utils/utils_log.c:249]: Maybe "Logger enabled (configuration)" or sth similar? Easier to understand.

File: scripts/docs_config/introduction.rst [hunk 37]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+By default, there is a guarantee that *error* messages are flushed immediately. One can change this behavior to flush on lower-level messages.
+
+Loggers redirect messages to *stdout*, *stderr*, or a file
+
+By default, no messages are printed. To enable logger you have to set **UMF_LOG** environment variable which have following syntax for setting logger options:
+
+  "[level:debug|info|warning|error];[flush:<debug|info|warning|error>];[output:stdout|stderr|file=<path>];[timestamp:yes|no][pid:yes|no]"
+
+  * level - a log level, meaning that only messages from this level and above are printed,
+            possible values, from the lowest level to the highest one: *debug*, *info*, *warning*, *error*,
```
Comment (bratpiorka) [inline scripts/docs_config/introduction.rst:113]: break this into two sentences: "... printed. Possible values, ... "

Comment (lplewa) [reply scripts/docs_config/introduction.rst:113]: Done

File: src/utils/utils_log.c (hunk 35 above)
Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: scripts/docs_config/introduction.rst (hunk 34 above)
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: done

File: scripts/docs_config/introduction.rst (hunk 33 above)
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: done

File: scripts/docs_config/introduction.rst (hunk 32 above)
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

File: scripts/docs_config/introduction.rst (hunk 34 above)
Comment (PatKamin) [reply scripts/docs_config/introduction.rst]: paramiter -> parameter

File: scripts/docs_config/introduction.rst [hunk 38]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+  * flush - a flush level, meaning that messages at this level and above are guaranteed to be flushed immediately,
+            possible values are the same as above,
+  * output - indicates where messages should be printed.
+             Possible values are: *stdout*, *stderr* and *file*,
+             when providing a *file* output option, a *<path>* is required
+  * timestamp - add a timestamp to log message
+  * pid - add thread and process ids to log message
+
+  .. note::
+    For output to file, a path to the file have to be provided after a coma, like in the example above. The path has to exist, file will be created if not existing. Path to the file should be not longer then 255 characters long.
```
Comment (PatKamin) [inline scripts/docs_config/introduction.rst:123]: comma

Comment (lplewa) [reply scripts/docs_config/introduction.rst]: done.

File: scripts/docs_config/introduction.rst (hunk 34 above)
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

**Review by PatKamin — APPROVED** [review]

File: scripts/docs_config/introduction.rst (hunk 37 above)
Comment (bratpiorka) [reply scripts/docs_config/introduction.rst:113]: done? I don't see the change

File: scripts/docs_config/introduction.rst [hunk 39]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+  * flush - a flush level, meaning that messages at this level and above are guaranteed to be flushed immediately,
+            possible values are the same as above,
+  * output - indicates where messages should be printed.
+             Possible values are: *stdout*, *stderr* and *file*,
+             when providing a *file* output option, a *<path>* is required
+  * timestamp - add a timestamp to log message
+  * pid - add thread and process ids to log message
+
+  .. note::
+    For output to file, a path to the file have to be provided after a comma, like in the example above. The path has to exist, file will be created if not existing. Path to the file should be not longer then 255 characters long.
```
Comment (igchor) [inline scripts/docs_config/introduction.rst:123]: then -> than

Comment (igchor) [inline scripts/docs_config/introduction.rst:123]: have -> has

**Review by igchor — APPROVED** [review]

File: src/utils/utils_log.c [hunk 40]
```diff
@@ -0,0 +1,251 @@
...
+        return "INFO";
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+        return "";
+    }
+}
+
+static const char *match_env_var(const char *envVar, const char *prefix,
```
Comment (igchor) [inline src/utils/utils_log.c:61]: Could you please move this function into utils_common? I think it would be useful for parsing other env vars as well (e.g. for customizing test behavior under qemu, etc.)

Comment (lplewa) [reply src/utils/utils_log.c:61]: Done

File: scripts/docs_config/introduction.rst (hunk 39 above)
Comment (lplewa) [reply scripts/docs_config/introduction.rst:123]: Done

Comment (lplewa) [reply scripts/docs_config/introduction.rst:123]: Done

File: scripts/docs_config/introduction.rst (hunk 37 above)
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Strange - now it's done.

File: test/test_utils_log.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (lplewa) [reply test/test_utils_log.cpp]: done

**Review by igchor — APPROVED** [review]

File: src/libumf_linux.c [hunk 41]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- * Copyright (C) 2024 Intel Corporation
+ * Copyright (C) 2023-2024 Intel Corporation
```
Comment (KFilipek) [inline src/libumf_linux.c:3]: Is it necessary? In `libumf_windows.c` there is just `2024`.
If it's desired please resolve my comment.

File: src/utils/utils_log.c [hunk 42]
```diff
@@ -0,0 +1,222 @@
...
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+        return "";
+    }
+}
+
+void util_log(util_log_level_t level, const char *format, ...) {
+
```
Comment (KFilipek) [inline src/utils/utils_log.c:64]: no needed line

Comment (lplewa) [reply src/utils/utils_log.c]: Done

File: src/libumf_linux.c (hunk 41 above)
Comment (lplewa) [reply src/libumf_linux.c]: Done

**Review by KFilipek — APPROVED** [review]
If this PR fixes 247, please change in the OP:
`ref: #247` -> `fixes #247`
Or pick the word you prefer: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: test/test_utils_log.cpp (hunk 29 above)
Comment (lukaszstolarczuk) [reply test/test_utils_log.cpp]: heh, the problem with unit tests is... that they are only unit tests :P We know both of these functionalities works independently, but we could have some higher-level tests as well. Nonetheless, it's ok for now, I guess.

File: src/utils/utils_common.h [hunk 43]
```diff
@@ -61,6 +61,23 @@ int util_env_var(const char *envvar, char *buffer, size_t buffer_size);
 // Check if the environment variable contains the given string.
 int util_env_var_has_str(const char *envvar, const char *str);
 
+// utils_parse_var - Parses var for a prefix,
```
[the remark this answers is not in the document]
Comment (lplewa) [reply src/utils/utils_common.h]: Done

File: scripts/docs_config/introduction.rst [hunk 44]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+  * output - indicates where messages should be printed.
+             Possible values are: *stdout*, *stderr* and *file*,
+             when providing a *file* output option, a *<path>* is required
+  * timestamp - add a timestamp to log messagex
+  * pid - add thread and process ids to log message
+
+  .. note::
+    For output to file, a path to the file has to be provided after a comma, like in the example above. The path has to exist, file will be created if not existing. Path to the file should be not longer than 255 characters long.
+    The output parameter is required, all other logger options are optional. The defaults are set when options are not provided in the environment variable.
+    Options have to be separated with `;`, option names, and their values with `:`. Additionally, when providing *file* output, the keyword *file* and a path to a file
```
[the remark this answers is not in the document]
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

[the remark this answers is not in the document]
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

File: scripts/docs_config/introduction.rst [hunk 45]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+Logging in UMF is handled by logger. There are several levels of logging: *debug*, *info*, *warning*, and *error*.
+The level of logging determines what messages will be printed, ie. the level set to *warning* means all messages at levels *warning* and *error* will be printed.
+
+By default, there is a guarantee that *error* messages are flushed immediately. One can change this behavior to flush on lower-level messages.
+
+Loggers redirect messages to *stdout*, *stderr*, or a file
+
+By default, no messages are printed. To enable logger you have to set **UMF_LOG** environment variable which have following syntax for setting logger options:
+
+  "[level:debug|info|warning|error];[flush:<debug|info|warning|error>];[output:stdout|stderr|file,<path>];[timestamp:yes|no];[pid:yes|no]"
```
[the remark this answers is not in the document]
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

File: scripts/docs_config/introduction.rst [hunk 46]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+  * flush - a flush level, meaning that messages at this level and above are guaranteed to be flushed immediately,
+            possible values are the same as above,
+  * output - indicates where messages should be printed.
+             Possible values are: *stdout*, *stderr* and *file*,
+             when providing a *file* output option, a *<path>* is required
+  * timestamp - add a timestamp to log messagex
+  * pid - add thread and process ids to log message
+
+  .. note::
+    For output to file, a path to the file has to be provided after a comma, like in the example above. The path has to exist, file will be created if not existing. Path to the file should be not longer than 255 characters long.
```
[the remark this answers is not in the document]
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

File: scripts/docs_config/introduction.rst [hunk 47]
```diff
@@ -95,6 +95,45 @@ the `allocation API`_ as a first argument. There is also a possibility to
...
+  "[level:debug|info|warning|error];[flush:<debug|info|warning|error>];[output:stdout|stderr|file,<path>];[timestamp:yes|no];[pid:yes|no]"
+
+  * level - a log level, meaning that only messages from this level and above are printed.
+            Possible values, from the lowest level to the highest one: *debug*, *info*, *warning*, *error*,
+  * flush - a flush level, meaning that messages at this level and above are guaranteed to be flushed immediately,
+            possible values are the same as above,
+  * output - indicates where messages should be printed.
+             Possible values are: *stdout*, *stderr* and *file*,
+             when providing a *file* output option, a *<path>* is required
+  * timestamp - add a timestamp to log messagex
```
[the remark this answers is not in the document]
Comment (lplewa) [reply scripts/docs_config/introduction.rst]: Done

**Review by lukaszstolarczuk — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]
